### PR TITLE
schedule watchTypeIfUnseen to prevent loop

### LINKED ISSUE #8006: loop(max stack size exceeded) if creating an record, whose type was not loaded yet. and ember-inspector open
createRecord will end up calling watchTypeIfUnseen ,
watchTypeIfUnseen is calling wrapModelType , which calls getRecords, which proceeds to do a sync, which then ends up calling watchTypeIfUnseen. See the stack strace in the description

### Reproduction

need to have ember-inspector open

this.store.createRecord('my-record', my-data).save();

### Description

-private.js:733 Uncaught RangeError: Maximum call stack size exceeded
    at get type [as type] (-private.js:733:1)
    at store._createRecordData (index.js:70:1)
    at InternalModel.get (-private.js:6136:1)
    at InternalModel._isRecordFullyDeleted (-private.js:5126:1)
    at InternalModel.isHiddenFromRecordArrays (-private.js:5120:1)
    at shouldIncludeInRecordArrays (-private.js:6936:1)
    at RecordArrayManager._visibleIdentifiersByType (-private.js:7095:1)
    at RecordArrayManager._syncLiveRecordArray (-private.js:7034:1)
    at RecordArrayManager.liveRecordArrayFor (-private.js:7076:1)
    at Store.peekAll (-private.js:9691:1)
<html>
<body>

  | function | @ | file
-- | -- | -- | --
  | get type | @ | -private.js:733
  | store._createRecordData | @ | index.js:70
  | get | @ | -private.js:6136
  | _isRecordFullyDeleted | @ | -private.js:5126
  | isHiddenFromRecordArrays | @ | -private.js:5120
  | shouldIncludeInRecordArrays | @ | -private.js:6936
  | _visibleIdentifiersByType | @ | -private.js:7095
  | _syncLiveRecordArray | @ | -private.js:7034
  | liveRecordArrayFor | @ | -private.js:7076
  | peekAll | @ | -private.js:9691
  | getRecords | @ | index.js:187
  | wrapModelType | @ | data_adapter.js:481
  | watchTypeIfUnseen | @ | index.js:111
  | store._createRecordData | @ | index.js:70
  | get | @ | -private.js:6136
  | _isRecordFullyDeleted | @ | -private.js:5126
  | isHiddenFromRecordArrays | @ | -private.js:5120
  | shouldIncludeInRecordArrays | @ | -private.js:6936
  | _visibleIdentifiersByType | @ | -private.js:7095
  | _syncLiveRecordArray | @ | -private.js:7034
  | liveRecordArrayFor | @ | -private.js:7076
  | peekAll | @ | -private.js:9691
  | getRecords | @ | index.js:187
  | wrapModelType | @ | data_adapter.js:481
  | watchTypeIfUnseen | @ | index.js:111
  | store._createRecordData | @ | index.js:70
  | get | @ | -private.js:6136
  | _isRecordFullyDeleted | @ | -private.js:5126
  | isHiddenFromRecordArrays | @ | -private.js:5120
  | shouldIncludeInRecordArrays | @ | -private.js:6936

</body>
</html>

### Versions

"ember-data": "^4.4.0",
"ember-source": "^4.4.0",

### PR DESCRIPTION
**Pull request by patricklx**
fixes #8006

### CODE REVIEW

**Comment by patricklx** [comment]
could also fix https://github.com/emberjs/ember-inspector/issues/1875

**Comment by kdagnan** [comment]
@patricklx I was having a similar issue & stack trace to you - but I just tried running your branch locally and it hasn't fixed anything for me. Funny enough it's on a regex call. Do you think you could look at this one? Might be a quick/similar fix. 
`Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at RegExp.exec (<anonymous>)
    at RegExp.test (<anonymous>)
    at isBlank (index.js:2729:1)
    at Object.isPresent (index.js:2768:1)
    at Store.modelFor (-private.js:10613:1)
    at Class.watchTypeIfUnseen (index.js:105:1)
    at store._createRecordData (index.js:69:1)
    at get _recordData [as _recordData] (-private.js:4989:1)
    at InternalModel._isRecordFullyDeleted (-private.js:5024:1)
    at InternalModel.isHiddenFromRecordArrays (-private.js:5018:1)`

**Comment by patricklx** [comment]
I think that's something else. You can get a Maximum call stack size exceeded with the regex.test alone as well if its on a strange string

**Comment by snewcomer** [comment]
@patricklx Would you mind adding a test if applicable to `save-test.js`?

**Comment by kdagnan** [comment]
Can confirm this fix addressed similar loop issues for me. Originally it did not because I was running ember 3.28, ember-data 4.4 (your branch), and ember-cli 4.2.0. Updating ember-cli to 4.4.0 and manually adding your fix to the NPM-installed ember-data 3.28.10 fixed this issue. 

Could we get this applied to a 3.28.11 as well? :)

**Comment by spruce** [comment]
This fixes #8006 for me

File: packages/-ember-data/tests/integration/debug-adapter-test.js [hunk 1]
```diff
@@ -39,31 +39,21 @@ if (has('@ember-data/debug')) {
...
       }
 
-      function updated(types) {
-        assert.strictEqual(types[0].count, 1, 'We updated one record');
-      }
-
-      debugAdapter.watchModelTypes(added, updated);
+      debugAdapter.watchModelTypes(added, () => null);
 
-      store.push({
```
Comment (runspired) [inline packages/-ember-data/tests/integration/debug-adapter-test.js:59]: we should keep a test for push as well

**Review by runspired — APPROVED** [review]

**Comment by runspired** [comment]
@patricklx looks like the tests fail, mind rebasing and looking into that?

**Comment by brandensilva** [comment]
I would also appreciate a 3.28.11 release on this one. It had me scratching my head for half a day wondering how my typescript upgrades were call stacking and triggering illegal character issues. Thanks for fixing this @patricklx!

**Comment by runspired** [comment]
@brandensilva in the works

**Pull request merged**